File: 06_breadth-first_search/ts/01_breadth-first_search.ts
interface Graph<T> {
  [key: string]: T;
}

const graph: Graph<string[]> = {};
graph["you"] = ["alice", "bob", "claire"];
graph["bob"] = ["anuj", "peggy"];
graph["alice"] = ["peggy"];
graph["claire"] = ["thom", "jonny"];
graph["anuj"] = [];
graph["peggy"] = [];
graph["thom"] = [];
graph["jonny"] = [];

const personIsSeller = (name: string): boolean => name.endsWith("m");

const search = (name: string): boolean => {
  let searchQueue: string[] = [...graph[name]];

  const searched: string[] = [];
  while (searchQueue.length > 0) {
    const person = searchQueue.shift();

    if (person && !searched.includes(person)) {
      if (personIsSeller(person)) {
        console.log(`${person} is a mango seller!`);
        return true;
      } else {
        searchQueue = searchQueue.concat(graph[person]); 
        searched.push(person);
      }
    }
  }
  return false;
};

search("you");

File: 06_breadth-first_search/scala/01_breadth-first_search.scala
import scala.collection.mutable

def personIsSeller(name: String): Boolean = {
  name.last == 'm'
}

val graph = mutable.Map[String, List[String]](
  "you" -> List("alice", "bob", "claire"),
  "bob" -> List("anuj", "peggy"),
  "alice" -> List("peggy"),
  "claire" -> List("thom", "jonny"),
  "anuj" -> List(),
  "peggy" -> List(),
  "thom" -> List(),
  "jonny" -> List()
)

def search(name: String): Boolean = {
  val searchQueue = mutable.Queue[String]()
  searchQueue.enqueue(name)
  // This is how you keep track of which people you've searched before.
  val searched = mutable.Set[String]()
  
  while(searchQueue.nonEmpty) {
    val person = searchQueue.dequeue()
    // Only search this person if you haven't already searched them.
    if(!searched.contains(person)) {
      if(personIsSeller(person)) {
        println(person + " is a mango seller!")
        return true
      }
      searchQueue.enqueueAll(graph(person))
      // Marks this person as searched
      searched.add(person)
    }
  }
  false
}

search("you")

